### PR TITLE
S3 dualstack mode check

### LINKED ISSUE #3720: Need a way to stop tempo from using dual stack urls for s3 backend
**Tempo compactor using s3 dualstack urls to interact with s3 backend storage, need a way to turn this behaviour off**
- Tempo compactor using s3 dual stack urls which are not supported by vpc private link services,  provided by aws. With this feature not available, the cost we are bearing due to transfer through network egress methods such as nat gateway are very high. We need a provision to reset this behaviour to the default manner of using the default s3 url which supports vpc endpoints

**To Reproduce**
Steps to reproduce the behaviour:
1. Start Tempo - 
tempo-distributed-1.9.1	2.4.1
2. Perform Operations (Read/Write/Others)

**Expected behaviour**
- There should be an option to control this behaviour of s3 upload and ensure that we can choose between dualstack endpoints and native endpoints

**Environment:**
 - Infrastructure: Kubernetes
 - Deployment tool: helm

### PR DESCRIPTION
**Pull request by sid-jar**
**What this PR does**: Adds a boolean param enabling (disabled by default) the use of dualstack mode on s3 which minio does by default

**Which issue(s) this PR fixes**:  
Fixes #3720

**Checklist**
- [ ] Tests updated
- [x] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Review by mapno — CHANGES_REQUESTED** [review]

**Review by sid-jar — COMMENTED** [review]
Made changes as suggested @mapno

File: docs/sources/tempo/configuration/manifest.md [hunk 1]
```diff
@@ -655,6 +655,7 @@ storage:
             hedge_requests_up_to: 2
             signature_v2: false
             forcepathstyle: false
+            enabledualstack: false
```
Comment (mapno) [inline docs/sources/tempo/configuration/manifest.md:658]:
```suggestion
            enable_dual_stack: false
```

File: docs/sources/tempo/configuration/manifest.md [hunk 2]
```diff
@@ -736,6 +737,7 @@ overrides:
                 hedge_requests_up_to: 2
                 signature_v2: false
                 forcepathstyle: false
+                enabledualstack: false
```
Comment (mapno) [inline docs/sources/tempo/configuration/manifest.md:740]:
```suggestion
                enable_dual_stack: false
```

File: docs/sources/tempo/setup/linux.md [hunk 3]
```diff
@@ -109,6 +109,7 @@ storage:
       endpoint: s3.us-east-1.amazonaws.com
       bucket: grafana-traces-data
       forcepathstyle: true
+      enabledualstack: false
```
Comment (mapno) [inline docs/sources/tempo/setup/linux.md:112]:
```suggestion
      enable_dual_stack: false
```

File: docs/sources/tempo/configuration/_index.md [hunk 4]
```diff
@@ -873,6 +873,10 @@ storage:
             # enable to use path-style requests.
             [forcepathstyle: <bool>]
 
+            # optional.
+            # enable to use dualstack endpoint for DNS resolution.
+            [enabledualstack: <bool>]
```
Comment (mapno) [inline docs/sources/tempo/configuration/_index.md:878]:
```suggestion
            [enable_dual_stack: <bool>]
```

File: docs/sources/tempo/configuration/_index.md [hunk 5]
```diff
@@ -873,6 +873,10 @@ storage:
             # enable to use path-style requests.
             [forcepathstyle: <bool>]
 
+            # optional.
+            # enable to use dualstack endpoint for DNS resolution.
```
Comment (mapno) [inline docs/sources/tempo/configuration/_index.md:877]: Could you link to S3 dualstack docs?

https://docs.aws.amazon.com/AmazonS3/latest/userguide/dual-stack-endpoints.html

**Review by mapno — CHANGES_REQUESTED** [review]
Left a few more comments. Could you add a changelog entry as well?

Also, you'll need to sign the CLA. Thanks!

**Comment by mapno** [comment]
Have you run `make vendor-check`? Pipeline is failing because of that.

**Comment by mapno** [comment]
Run `make update-mod` instead, my bad.

**Comment by sid-jar** [comment]
@mapno I've made the necessary changes. Let me know if it looks good

File: docs/sources/tempo/configuration/_index.md (hunk 4 above)
Comment (mapno) [reply docs/sources/tempo/configuration/_index.md:879]:
```suggestion
            [enable_dual_stack: <bool>]
```

**Comment by mapno** [comment]
You need to sign the CLA. It's in the first comment https://github.com/grafana/tempo/pull/3721#issuecomment-2135011570. Thanks

File: tempodb/backend/s3/s3.go [hunk 6]
```diff
@@ -660,6 +660,8 @@ func createCore(cfg *Config, hedge bool) (*minio.Core, error) {
 		opts.BucketLookup = minio.BucketLookupType(cfg.BucketLookupType)
 	}
 
+	opts.UseDualStack = cfg.UseDualStack
```
Comment (mapno) [inline tempodb/backend/s3/s3.go:663]: This option does not exist. Disabling dualstack endpoints is configured via a method in the client https://github.com/minio/minio-go/pull/1945/files#diff-f7bb71ca856c07b3329265f26b1e577e00b71e4792550cf14346bdbe0e7bd18dR342-R346.

Comment (mapno) [reply tempodb/backend/s3/s3.go:663]: This is why the pipeline is failing

**Comment by mapno** [comment]
Although not mandatory in this case, a unit test would be appreciated.

File: tempodb/backend/s3/s3.go (hunk 6 above)
Comment (sid-jar) [reply tempodb/backend/s3/s3.go:663]: So in this case, according to [this](https://arc.net/l/quote/vhuzymsw) function, the minio is going to use the dualstack endpoint everytime there's an s3 url present, right ?. How can we override this ?

Comment (mapno) [reply tempodb/backend/s3/s3.go:663]: You just need to call that function with `false`

**Comment by sid-jar** [comment]
@mapno , I've made the suggested changes. Does it look fine now ?

File: tempodb/backend/s3/s3.go [hunk 7]
```diff
@@ -660,6 +660,12 @@ func createCore(cfg *Config, hedge bool) (*minio.Core, error) {
 		opts.BucketLookup = minio.BucketLookupType(cfg.BucketLookupType)
 	}
 
+	clnt := &minio.Client{}
+
+	if cfg.UseDualStack {
+		clnt.SetS3EnableDualstack(false)
+	}
```
Comment (mapno) [inline tempodb/backend/s3/s3.go:667]: This client is never returned. You need to apply the config to the client returned by this function `newCore()`.

Comment (sid-jar) [reply tempodb/backend/s3/s3.go]: @mapno , I have modified the core and api logic. But `make vendor-check` both on local and on the workflow deletes the changes made. Can you help me here

File: tempodb/backend/s3/s3.go [hunk 8]
```diff
@@ -648,10 +648,11 @@ func createCore(cfg *Config, hedge bool) (*minio.Core, error) {
...
 	opts := &minio.Options{
-		Region:    cfg.Region,
-		Secure:    !cfg.Insecure,
-		Creds:     creds,
-		Transport: transport,
+		Region:           cfg.Region,
+		Secure:           !cfg.Insecure,
+		Creds:            creds,
+		Transport:        transport,
+		DualStackEnabled: cfg.UseDualStack,
```
Comment (mapno) [inline tempodb/backend/s3/s3.go:655]: This option doesn't exist, it's configured via a method [`SetS3EnableDualstack`](https://github.com/minio/minio-go/blob/14b3aa6d56708fdf2071346aadcdb085069d8937/api.go#L342-L346) on the client (what's returned by `minio.NewCore()`)

**Review by mapno — CHANGES_REQUESTED** [review]
I appreciate you might not be familiar with Go or Tempo's codebase, but please stop pushing broken commits. The [contributing guide](https://github.com/grafana/tempo/blob/main/CONTRIBUTING.md) I initially linked describes how you can make sure that the build at least compiles and passes the tests.

If you don't understand the change that you need to do, please let us know and we'd be glad to help.

File: tempodb/backend/s3/s3.go (hunk 8 above)
Comment (sid-jar) [reply tempodb/backend/s3/s3.go:655]: Sure @mapno , I appreciate your help throughout and would be glad if you're able to find time to fix this. I'll try to learn better and contribute. Thanks

**Review by mapno — CHANGES_REQUESTED** [review]
We're very close now

**Comment by sid-jar** [comment]
the `make vendor-check` still seems to fail. not sure how to get through that

**Comment by mapno** [comment]
You can run `make update-mod`

**Comment by sid-jar** [comment]
> You can run `make update-mod`

Yes, I did update-mode followed by a vendor-check. I assume its some sort of a warning and not any error from what I see.

`make: Entering directory '/src/loki'
goyacc -o pkg/traceql/expr.y.go pkg/traceql/expr.y && rm y.output

conflicts: 1 shift/reduce
make: Leaving directory '/src/loki'
go run ./pkg/parquetquerygen/predicates.go > ./pkg/parquetquery/predicates.gen.go
git diff --exit-code -- **/go.sum **/go.mod vendor/ pkg/tempopb/ pkg/traceql/`

**Comment by sikriJar** [comment]
I'm facing a similar problem. Hope this gets resolved soon! @mapno

**Comment by sid-jar** [comment]
@mapno , Can you guide me on what went wrong on the last pipeline. I can quickly make further changes as required..

**Comment by rohan-changejar** [comment]
Hi @mapno , if this pr is not upto your expectations, I can raise a new pr as well. Please let me know

File: docs/sources/tempo/configuration/_index.md (hunk 4 above)
Comment (mapno) [reply docs/sources/tempo/configuration/_index.md:879]: This needs changing still.

**Review by mapno — COMMENTED** [review]
I was away last week. The PR is now close to being approvable, there is only one comment left. Also, could you add a changelog entry?

My only worry is if the update is going to mess with S3 backends, now that the expiration for the credentials is being honoured. I suppose that's the intended behaviour, so that's OK.

File: docs/sources/tempo/configuration/_index.md [hunk 9]
```diff
@@ -873,6 +873,11 @@ storage:
             # enable to use path-style requests.
             [forcepathstyle: <bool>]
 
+            # optional.
```
Comment (knylander-grafana) [inline docs/sources/tempo/configuration/_index.md:876]:
```suggestion
            # Optional.
```

File: docs/sources/tempo/configuration/_index.md [hunk 10]
```diff
@@ -873,6 +873,11 @@ storage:
...
+            # enable to use dualstack endpoint for DNS resolution.
```
Comment (knylander-grafana) [inline docs/sources/tempo/configuration/_index.md:877]:
```suggestion
            # Optional.
            # Enable to use dualstack endpoint for DNS resolution. 
```

**Review by knylander-grafana — COMMENTED** [review]
Thank you for updating the documentation.

**Review by mapno — APPROVED** [review]
LGTM

**Comment by sid-jar** [comment]
Thanks @mapno, @knylander-grafana for the approval and guidance. One more thing, I don't see a way to draft a release (as mentioned on the releases.md), so I assume its done from the codeowners side. Is there a time bound on when can we expect a new release ?

**Comment by mapno** [comment]
>Is there a time bound on when can we expect a new release as this alleviates a huge cost infliction on our systems?

We aim to do a release roughly once every quarter (3 months). We released v2.5 less then a month ago, so I wouldn't expect to have v2.6 before a couple of months.

In the meantime, we publish builds for all commits in `main` in [DockerHub](https://hub.docker.com/r/grafana/tempo/tags).

**Comment by sid-jar** [comment]
@mapno thanks for the suggestion, could you please guide me on how to use this version on my helm chart. We tried replacing the docker version with the commit tag from docker hub. This didn't work for us.

**Pull request merged**